### PR TITLE
feat: quadratic + cubic sampling

### PR DESCRIPTION
**Pull request by AlpinDale**
This PR adds Quadratic and Cubic Sampling methods to vLLM.
Ref: oobabooga/text-generation-webui#5403 and oobabooga/text-generation-webui#5551

### CODE REVIEW

File: vllm/model_executor/sampling_metadata.py [hunk 1]
```diff
@@ -228,6 +254,10 @@ def from_lists(cls, temperatures: List[float], top_ps: List[float],
             top_ps=top_ps_t.to(device=device, non_blocking=True),
             top_ks=top_ks_t.to(device=device, non_blocking=True),
             min_ps=min_ps_t.to(device=device, non_blocking=True),
+            smoothing_factors=smoothing_factors_t.to(device=device,
+                                                     non_blocking=True),
+            smoothing_curve=smoothing_curves_t.to(device=device,
```
Comment (Yard1) [inline vllm/model_executor/sampling_metadata.py:259]:
```suggestion
            smoothing_curves=smoothing_curves_t.to(device=device,
```

File: vllm/model_executor/layers/sampler.py [hunk 2]
```diff
@@ -242,6 +244,30 @@ def _apply_min_p(
     return logits
 
 
+def _apply_quadratic_sampling(
```
Comment (Yard1) [inline vllm/model_executor/layers/sampler.py:247]:
```suggestion
# torch.jit will fuse pointwise ops for better performance
@torch.jit.script
def _apply_quadratic_sampling(
```

**Comment by Yard1** [comment]
@AlpinDale Thanks, this looks good to me. Could we add a test?

**Comment by AlpinDale** [comment]
Thanks for the review @Yard1. It needs a bit more work, could probably use a mask and an unsqueeze for the factor/curve. I'll finish this up soon and add a test unit.

File: vllm/model_executor/sampling_metadata.py [hunk 3]
```diff
@@ -228,6 +254,10 @@ def from_lists(cls, temperatures: List[float], top_ps: List[float],
             top_ps=top_ps_t.to(device=device, non_blocking=True),
             top_ks=top_ks_t.to(device=device, non_blocking=True),
             min_ps=min_ps_t.to(device=device, non_blocking=True),
+            smoothing_factors=smoothing_factors_t.to(device=device,
+                                                     non_blocking=True),
+            smoothing_curves=smoothing_curves_t.to(device=device,
+                                                  non_blocking=True),
```
Comment (Yard1) [inline vllm/model_executor/sampling_metadata.py:260]:
```suggestion
            smoothing_curves=smoothing_curves_t.to(device=device,
                                                   non_blocking=True),
```

**Comment by AlpinDale** [comment]
@Yard1 Should be done, let me know if there's any issues. Correctness tests pass for the new sampler.

File: vllm/model_executor/layers/sampler.py [hunk 4]
```diff
@@ -242,6 +247,42 @@ def _apply_min_p(
...
+    max_logits = logits.max(dim=-1, keepdim=True).values
+    diff = logits - max_logits
+    smoothing_factors.unsqueeze_(dim=1)
+    smoothing_curves.unsqueeze_(dim=1)
+
+    k = (3 - smoothing_curves) / 2
+    s = (smoothing_curves - 1) / 2
+
+    mask = smoothing_factors > 0
+    mask = mask.expand(*logits.shape)
```
Comment (Yard1) [inline vllm/model_executor/layers/sampler.py:275]:
```suggestion
    mask = mask.expand(logits.shape[0], logits.shape[1])
```

File: vllm/model_executor/layers/sampler.py [hunk 5]
```diff
@@ -242,6 +247,42 @@ def _apply_min_p(
...
+    k = (3 - smoothing_curves) / 2
+    s = (smoothing_curves - 1) / 2
+
+    mask = smoothing_factors > 0
+    mask = mask.expand(logits.shape[0], logits.shape[1])
+
+    # only transform logits when they're not -inf, otherwise
+    # fails at smoothing_curves==3
+    transformed_logits = torch.where(
+        (logits != float('-inf')) & mask, -(k * smoothing_factors * diff**2) +
```
Comment (Yard1) [inline vllm/model_executor/layers/sampler.py:280]:
```suggestion
        (logits != float('-inf')) & ~mask, -(k * smoothing_factors * diff**2) +
```

File: vllm/model_executor/layers/sampler.py [hunk 6]
```diff
@@ -242,6 +247,42 @@ def _apply_min_p(
...
+    k = (3 - smoothing_curves) / 2
+    s = (smoothing_curves - 1) / 2
+
+    mask = smoothing_factors > 0
+    mask = mask.expand(logits.shape[0], logits.shape[1])
+
+    # only transform logits when they're not -inf, otherwise
+    # fails at smoothing_curves==3
+    transformed_logits = torch.where(
+        (logits != float('-inf')) & ~mask, -(k * smoothing_factors * diff**2) +
```
Comment (Yard1) [inline vllm/model_executor/layers/sampler.py:280]:
```suggestion
        (logits != float('-inf')) & mask, -(k * smoothing_factors * diff**2) +
```

**Comment by Yard1** [comment]
Hey @AlpinDale there seems to be an error in the test (the condition is not being met) that was hidden before as `logits` was modified in place. Could you take a look?

**Comment by hmellor** [comment]
@AlpinDale do you plan to continue this line of work?

**Comment by AlpinDale** [comment]
> @AlpinDale do you plan to continue this line of work?

This sampler is quite outdated by now, so I'm not sure it'll be a useful addition anymore. There are better alternatives.

**Comment by hmellor** [comment]
Ok, thank you for the update!
